### PR TITLE
Add system test for hostless mode.

### PR DESCRIPTION
**Pull request by o01eg**
Improvement for #2050. Should catch errors like #1897.

Test acts as a player on multiplayer server, manages by environment variables:
- Do start a server with hostless mode if `FO_TEST_HOSTLESS_LAUNCH_SERVER` was set with save enabled if `FO_TEST_HOSTLESS_SAVE` was set.
- Do connect to lobby to localhost server as a Player.
- Expect successfully connection to localhost server.
- Do add `FO_TEST_HOSTLESS_AIS` AIs to lobby (by default 2).
- Expect AIs will be added successfully.
- Do make player ready.
- Expect game started.
- Do make empty turns until turn number reach `FO_TEST_HOSTLESS_TURNS` or mock player will be eliminated or someone will win.
- Expect turns're processing without error.
- Do disconnect from server.
- Do reconnect to server until number of played games reach `FO_TEST_HOSTLESS_GAMES`.
- Expect got to lobby.
- Expect number of AIs was preserved.
- Do repeat game steps.
- Expect all games processed correctly.
- Do shut down server.
- Expect that server and AIs were shut down.

### CODE REVIEW

File: test/system/ClientAppFixture.cpp [hunk 1]
```diff
@@ -37,6 +39,12 @@ bool ClientAppFixture::PingLocalHostServer()
 bool ClientAppFixture::ConnectToLocalHostServer()
 { return m_networking->ConnectToLocalHostServer(); }
 
+bool ClientAppFixture::ConnectToServer(const std::string& ip_address)
+{ return m_networking->ConnectToServer(ip_address); }
+
+void ClientAppFixture::DisconnectFromServer()
+{ return m_networking->DisconnectFromServer();  }
```
Comment (geoffthemedio) [inline test/system/ClientAppFixture.cpp]: extra space

File: test/system/SmokeTestHostless.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+                UpdateLobby();
+                start_time = boost::posix_time::microsec_clock::local_time();
+                while (!m_lobby_updated) {
+                    BOOST_REQUIRE(ProcessMessages(start_time, MAX_WAITING_SEC));
+                }
+            }
+            BOOST_REQUIRE(GetLobbyAICount() == num_AIs);
+
+        } else {
+            // other game should retain AI from previous game
```
Comment (geoffthemedio) [inline test/system/SmokeTestHostless.cpp]: retain AI clients or retain number of AIs to start up?

**Comment by geoffthemedio** [comment]
otherwise no comments... no idea if it works but I assume you've tested the testing.

File: test/system/SmokeTestHostless.cpp (hunk 2 above)
Comment (o01eg) [reply test/system/SmokeTestHostless.cpp]: Number of AI clients. Fixed.

File: test/system/SmokeTestHostless.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        args.push_back("/proc/self/fd/1");
+#endif
+
+        server = Process(SERVER_CLIENT_EXE, args);
+
+        BOOST_TEST_MESSAGE("Server started.");
+    }
+
+    for (unsigned int g = 0; g < num_games; ++g) {
+        BOOST_TEST_MESSAGE(g << "Game. Connecting to server...");
```
[the remark this answers is not in the document]
Comment (o01eg) [reply test/system/SmokeTestHostless.cpp]: Fixed.

File: test/system/ClientAppFixture.cpp (hunk 1 above)
Comment (o01eg) [reply test/system/ClientAppFixture.cpp]: Fixed.

**Comment by o01eg** [comment]
Yep, Travis CI successfully executed this test. And with more turns and games it even reproduced #1897.

**Pull request merged**